keps/sig-apps/2307-job-tracking-wihout-lingering-pods/README.md

```diff
@@ -0,0 +1,573 @@
+relies on the garbage collector having a big threshold.
+
+### Goals
+
+- Perform Job completion and failures tracking without relying on lingering
+  Pods.
+
+### Non-Goals
+
+- Remove Pods once they have been accounted for.
+
+## Proposal
+
+The Job controller creates Pods with a finalizer to prevent finished Pods from
+being removed by the garbage collector. The Job controller removes the finalizer
+from the finished Pods once it has accounted for them. In subsequent Job syncs,
+the controller ignores finished Pods that don't have the finalizer.
+
+### Notes/Constraints/Caveats (Optional)
+
+Due to the lack of support of atomic changes across kubernetes objects, an
+intermediate state is necessary. Before removing the Pod finalizers,
+the controller adds finished Pods to a list in the Job status. After removing
+the Pod finalizers, the controller clears the list and updates the counters.
+
+### Risks and Mitigations
+
+#### New API calls
+
+The new algorithm introduces new API calls:
+- one per Pod lifecycle to remove finalizers
+- one for each Job sync, to track the intermediate state.
+
+Note that there no new calls in the Pod creation path.
+
+On the other hand, to update the Job status once a Pod finishes, we need a total
+2 new API calls compared to the legacy algorithm. If more than one Pod finish at
+a given time, we add `n + 1` API calls, where `n` is the number finished Pods.
+
+This increase in API calls is justified for the following reasons:
+- The legacy tracking cannot handle a big number of terminated Pods, across
+  any number of Jobs, at a given point.
+- In the entirety of its lifecycle, a Pod requires at least 8 API calls,
+  including status updates and events.
+  
+However, in order to prevent Jobs with big number of Pods from starving Jobs
+with fewer Pods, the Job controller might skip status updates until enough
+Pods have accumulated or enough time has passed. See [Algorithm](#algorithm)
+below for more details.
+   
+#### Bigger Job status
+
+Job status can temporarily grow if too many Pods finish at the same time or
+if the Job controller is down for some time. In this case, we do partial
+status updates. See [Algorithm](#algorithm) below.
+
+#### Unprotected Job status endpoint
+
+Changes in the status not produced by the Job controller in
+kube-controller-manager could affect the Job tracking. Cluster administrators
+should make sure to protect the Job status endpoint via RBAC.
+
+## Design Details
+
+### API changes
+
+The Job status gets a new struct to hold the uncounted Pods before they are
+added to the counters.
+
+```golang
+type JobStatus struct {
+    Succeeded int32
+    Failed    int32
+    ...
+
+    // UncountedTerminatedPods holds UIDs of Pods that have finished but
+    // haven't been accounted in the counters.
+    // If nil, Job tracking doesn't make use of this structure.
+    // +optional
+    UncountedTerminatedPods *UncountedTerminatedPods
```

Comment:
@lavalamp @erictune - IIUC, in the past we were using a convention that status contains information, than can be computed based on the cluster state.
This is more or less, why things like "AcquireTime" or "RenewTime" are in LeaseSpec:
https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/api/coordination/v1/types.go#L40
[that was Brian`s suggestion back then].

Did we give up with this rule? If not, then all `Succeeded`, `Failed`, and `UncountedTerminatedPods` technically no longer belong to status...

[I don't have strong opinion here - just want to flag this for discussion.]

@liggitt - for his API reviewer/approver opinion

Reply:
Actually this makes the Success / Finished numbers un-reproducible from the environment. UncountedTerminatedPods is (at least partially) reproducible! I think we are just giving up on this criteria to make this work. It's worth a discussion but I don't see other options. The reproducibility criterion is (IMO) mostly aesthetic (we've never made the system do anything that relied on it and it seems quite unlikely that we will in the future) and it's not worth it to keep it for this object.

Reply:
> Actually this makes the Success / Finished numbers un-reproducible from the environment. UncountedTerminatedPods is (at least partially) reproducible! I think we are just giving up on this criteria to make this work. It's worth a discussion but I don't see other options. The reproducibility criterion is (IMO) mostly aesthetic (we've never made the system do anything that relied on it and it seems quite unlikely that we will in the future) and it's not worth it to keep it for this object.

agreed.

Reply:
AIUI, the intent of this API Convention was that status could be stored in some less reliable storage for performance.  5+ years on from this intention, we haven't made use of that possibility.  And I really doubt that all clients are prepared for the event where status disappears and then is later reconstructed.  I am thinking of e.g. kubectl waiting for a to be ready, and then all the status disappears.  Is the client code going to do the right thing?  What is the right thing?  We never designed for that.  So we should remove that API convention.

Reply:
Thanks - for me it was also kind of aesthetic thing, but I wanted to ensure that there is a general agreement on that.
It seems there is, which is what I wanted to achieve.

Reply:
I think the intent is also that this rule forces requirements on the cluster introspection mechanism to be able to surface anything important enough to track in status. If we have the ability to reconstruct status, then we know resources are not leaked out; our reconciliation is fully closed, without drift. If we put something in status that we cannot discover from the cluster itself we no longer have that guarantee - status could diverge from reality.

That said, in the case of job tracking, there are no resources leaked or lost track of. We lose track of some job stats potentially, but not things that are consuming underlying resources. So, only issue is the slippery slope.